From 852b162c8255b2bdecde76e89e272149548e0669 Mon Sep 17 00:00:00 2001
From: Sergey Dyasli <sergey.dyasli@citrix.com>
Date: Wed, 7 Nov 2018 09:31:17 +0000
Subject: [PATCH] Allocate space in structs pre-emptively to increase chances
 of livepatching

1. Expand FEATURESET for 4 more words
2. Increment CPUID_GUEST_NR_FEAT to get an additional feature leaf
   in struct cpuid_policy
3. Reserve space in the following structs:

    grant_table
    vtd_iommu
    amd_iommu
    guest_iommu
    hvm_ioreq_vcpu
    hvm_ioreq_server
    msr_domain_policy
    msr_vcpu_policy
    p2m_domain

Signed-off-by: Sergey Dyasli <sergey.dyasli@citrix.com>
diff --git a/xen/common/grant_table.c b/xen/common/grant_table.c
index 88c5ee24e8..964b7acf44 100644
--- a/xen/common/grant_table.c
+++ b/xen/common/grant_table.c
@@ -81,6 +81,9 @@ struct grant_table {
     const struct domain *domain;
 
     struct grant_table_arch arch;
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[2];
 };
 
 unsigned int __read_mostly opt_max_grant_frames = 64;
diff --git a/xen/drivers/passthrough/vtd/iommu.h b/xen/drivers/passthrough/vtd/iommu.h
index 216791b3d6..b70779c43c 100644
--- a/xen/drivers/passthrough/vtd/iommu.h
+++ b/xen/drivers/passthrough/vtd/iommu.h
@@ -540,6 +540,9 @@ struct vtd_iommu {
     struct list_head ats_devices;
     unsigned long *domid_bitmap;  /* domain id bitmap */
     u16 *domid_map;               /* domain id mapping array */
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[2];
 };
 
 #define INTEL_IOMMU_DEBUG(fmt, args...) \
diff --git a/xen/include/asm-x86/amd-iommu.h b/xen/include/asm-x86/amd-iommu.h
index 951a012565..38506fea91 100644
--- a/xen/include/asm-x86/amd-iommu.h
+++ b/xen/include/asm-x86/amd-iommu.h
@@ -103,6 +103,9 @@ struct amd_iommu {
     int enabled;
 
     struct list_head ats_devices;
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[2];
 };
 
 struct ivrs_mappings {
@@ -181,6 +184,9 @@ struct guest_iommu {
 
     /* guest interrupt settings */
     struct guest_iommu_msi  msi;
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[2];
 };
 
 extern bool_t iommuv2_enabled;
diff --git a/xen/include/asm-x86/hvm/domain.h b/xen/include/asm-x86/hvm/domain.h
index bcc5621797..1e415b5c3f 100644
--- a/xen/include/asm-x86/hvm/domain.h
+++ b/xen/include/asm-x86/hvm/domain.h
@@ -46,6 +46,9 @@ struct hvm_ioreq_vcpu {
     struct vcpu      *vcpu;
     evtchn_port_t    ioreq_evtchn;
     bool             pending;
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[1];
 };
 
 #define NR_IO_RANGE_TYPES (XEN_DMOP_IO_RANGE_PCI + 1)
@@ -67,6 +70,9 @@ struct hvm_ioreq_server {
     struct rangeset        *range[NR_IO_RANGE_TYPES];
     bool                   enabled;
     uint8_t                bufioreq_handling;
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[2];
 };
 
 /*
diff --git a/xen/include/asm-x86/msr.h b/xen/include/asm-x86/msr.h
index 4b8fd6ec53..0196a515bf 100644
--- a/xen/include/asm-x86/msr.h
+++ b/xen/include/asm-x86/msr.h
@@ -325,6 +325,9 @@ struct vcpu_msrs
      * values here may be stale in current context.
      */
     uint32_t dr_mask[4];
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[2];
 };
 
 void init_guest_msr_policy(void);
diff --git a/xen/include/asm-x86/p2m.h b/xen/include/asm-x86/p2m.h
index 94285db1b4..927a19be35 100644
--- a/xen/include/asm-x86/p2m.h
+++ b/xen/include/asm-x86/p2m.h
@@ -362,6 +362,9 @@ struct p2m_domain {
          unsigned int flags;
          unsigned long entry_count;
      } ioreq;
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[2];
 };
 
 /* get host p2m table */
diff --git a/xen/include/xen/lib/x86/cpuid.h b/xen/include/xen/lib/x86/cpuid.h
index df5946b6b1..d4f94e065a 100644
--- a/xen/include/xen/lib/x86/cpuid.h
+++ b/xen/include/xen/lib/x86/cpuid.h
@@ -80,7 +80,7 @@ const char *x86_cpuid_vendor_to_str(unsigned int vendor);
 
 #define CPUID_GUEST_NR_BASIC      (0xdu + 1)
 #define CPUID_GUEST_NR_CACHE      (5u + 1)
-#define CPUID_GUEST_NR_FEAT       (1u + 1)
+#define CPUID_GUEST_NR_FEAT       (2u + 1)
 #define CPUID_GUEST_NR_TOPO       (1u + 1)
 #define CPUID_GUEST_NR_XSTATE     (62u + 1)
 #define CPUID_GUEST_NR_EXTD_INTEL (0x8u + 1)
@@ -278,6 +278,8 @@ struct cpuid_policy
 static inline void cpuid_policy_to_featureset(
     const struct cpuid_policy *p, uint32_t fs[FEATURESET_NR_ENTRIES])
 {
+    unsigned int i;
+
     fs[FEATURESET_1d]  = p->basic._1d;
     fs[FEATURESET_1c]  = p->basic._1c;
     fs[FEATURESET_e1d] = p->extd.e1d;
@@ -289,6 +291,10 @@ static inline void cpuid_policy_to_featureset(
     fs[FEATURESET_e8b] = p->extd.e8b;
     fs[FEATURESET_7d0] = p->feat._7d0;
     fs[FEATURESET_7a1] = p->feat._7a1;
+
+    /* Zero reserved featureset words */
+    for ( i = FEATURESET_7a1 + 1; i < FEATURESET_NR_ENTRIES; i++ )
+        fs[i] = 0;
 }
 
 /* Fill in a CPUID policy from a featureset bitmap. */
diff --git a/xen/include/xen/lib/x86/msr.h b/xen/include/xen/lib/x86/msr.h
index e83a8fbb0f..04a95919a5 100644
--- a/xen/include/xen/lib/x86/msr.h
+++ b/xen/include/xen/lib/x86/msr.h
@@ -23,6 +23,9 @@ struct msr_policy
             bool cpuid_faulting:1;
         };
     } platform_info;
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[2];
 };
 
 #ifdef __XEN__
diff --git a/xen/tools/gen-cpuid.py b/xen/tools/gen-cpuid.py
index f76e80d690..d0cc13adec 100755
--- a/xen/tools/gen-cpuid.py
+++ b/xen/tools/gen-cpuid.py
@@ -130,6 +130,8 @@ def crunch_numbers(state):
 
     # Size of bitmaps
     state.nr_entries = nr_entries = (max(state.names.keys()) >> 5) + 1
+    reserved_entries_num = 4
+    state.nr_entries = nr_entries = nr_entries + reserved_entries_num
 
     # Features common between 1d and e1d.
     common_1d = (FPU, VME, DE, PSE, TSC, MSR, PAE, MCE, CX8, APIC,
